feat: update templates to specify control and worker shapes separately (#19)
oracle/cluster-api-provider-oci: docs/src/gs/create-workload-cluster.md

| `CONTROL_PLANE_MACHINE_COUNT` | 1              | The number of control plane machines for the workload cluster.                                                                                                                |
| `KUBERNETES_VERSION`          |                | The Kubernetes version to use for the workload cluster. If unspecified, the value from OS environment variables or the .cluster-api/clusterctl.yaml config file will be used. |
| `NAMESPACE`                   |                | The namespace for the workload cluster. If not specified, the current namespace is used.                                                                                      |
| `POD_CIDR`                    | 192.168.0.0/16 | CIDR range of the Kubernetes pod-to-pod network.                                                                                                                              |
| `SERVICE_CIDR`                | 10.128.0.0/12  | CIDR range of the Kubernetes pod-to-services network.                                                                                                                         |

> | `SERVICE_DOMAIN`              |                |                                                                                                                                                                               |

Reviewer: Why is there no description for this value? I didn't notice this in the previous review, but we shouldn't leave this undefined. If this is referring to the DNS domain name, it would be better to use `SERVICE_FQDN`, i.e. fully qualified domain name. 

Either way, it should have a description.

Reviewer: The description still doesn't actually tell me what this field is for or what it actually does. It's just the name of the field in prose format.

Reviewer 2: @joekr please remove this field fromt he templates and description, other providers (aws/azure) dont have the field in their templates and from a look at the code, the field is not the right place to override Kubernetes service domain name.

## Workload Cluster Parameters

The following Oracle Cloud Infrastructure (OCI) configuration parameters are available when creating a workload 
cluster on OCI using one of our predefined templates:

| Parameter                                 | Default Value       | Description                                                                                                                                                                                                                                                                                                                                                                                                  |
|-------------------------------------------|---------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `OCI_COMPARTMENT_ID`                      |                     | The OCID of the compartment where the OCI resources in which the resources are created                                                                                                                                                                                                                                                                                                                       |

Reviewer: ```suggestion
| `OCI_COMPARTMENT_ID`                      |                     | The OCID of the compartment in which to create the required compute, storage and network resources |
```

> | `OCI_IMAGE_ID`                            |                     | The OCID of the image (Oracle Linux or <br/>Ubuntu) with which to create the Kubernetes nodes. This same image is used for both the control plane and the worker nodes.                                                                                                                                                                                                                                      |

Reviewer 2: Dont think we need to specify Ol /Ubunt etc, just say image of kubernetes nodes. We may add more types later and we don't need to change the deocs every time.

> | `OCI_CONTROL_PLANE_SHAPE`                 | VM.Standard.E4.Flex | The shape of the Kubernetes nodes                                                                                                                                                                                                                                                                                                                                                                            |

Reviewer 2: shape of kubernetes control plane nodes?

| `OCI_CONTROL_PLANE_SHAPE_OCPUS`           | 1                   | The number of OCPUs allocated to the control plane instance                                                                                                                                                                                                                                                                                                                                                  |
| `OCI_WORKER_SHAPE`                        | VM.Standard.E4.Flex | The shape of the Kubernetes worker nodes                                                                                                                                                                                                                                                                                                                                                                     |

> | `OCI_WORKER_SHAPE_OCPUS`                  | 1                   | The number of OCPUs allocated to the worker instance                                                                                                                                                                                                                                                                                                                                                         |

Reviewer 2: can we add a reference to oci docs here?

Reviewer 2: for flex shapes

| `OCI_SSH_KEY`                             |                     | The public SSH key to be added to the Kubernetes nodes. It can be used to login to the node and troubleshoot failures.                                                                                                                                                                                                                                                                                       |

> | `OCI_CONTROL_PLANE_PV_TRANSIT_ENCRYPTION` | true                | [In-transit encryption](https://docs.oracle.<br/>com/en-us/iaas/Content/File/Tasks/intransitencryption.htm) provides a way to secure your data between instances and mounted file systems using TLS 1.2 (Transport Layer Security) encryption. Only [some bare metal instances](https://docs.oracle.com/en-us/iaas/releasenotes/changes/60d602f5-abb3-4639-aa19-292a5744a808/) support in-transit encryption |

Reviewer: ```suggestion
| `OCI_CONTROL_PLANE_PV_TRANSIT_ENCRYPTION` | true                | Enables [in-flight Transport Layer Security (TLS) 1.2 encryption](https://docs.oracle.com/en-us/iaas/Content/File/Tasks/intransitencryption.htm) of data between control plane nodes and their associated block storage devices |
```

I would then add a footnote to this: "Note that only [specific bare metal shapes](https://docs.oracle.com/en-us/iaas/releasenotes/changes/60d602f5-abb3-4639-aa19-292a5744a808/) support in-transit encryption. If an unsupported shape is specified, the deployment will..."

And then add either "fail completely" or "succeed with encryption disabled" based on the outcome of either an existing or future test to determine what the actual outcome is. I would suggest we go with pre-flight validation of the chosen shape with selected features and fast-fail on a mismatch.

> | `OCI_WORKER_PV_TRANSIT_ENCRYPTION`        | true                | [In-transit encryption](https://docs.oracle.com/en-us/iaas/Content/File/Tasks/intransitencryption.htm) provides a way to secure your data between instances and mounted file systems using TLS 1.2 (Transport Layer Security) encryption. Only [some bare metal instances](https://docs.oracle.com/en-us/iaas/releasenotes/changes/60d602f5-abb3-4639-aa19-292a5744a808/) support in-transit encryption      |

Reviewer: ```suggestion
| `OCI_WORKER_PV_TRANSIT_ENCRYPTION`        | true                | Enables [in-flight Transport Layer Security (TLS) 1.2 encryption](https://docs.oracle.com/en-us/iaas/Content/File/Tasks/intransitencryption.htm) of data between worker nodes and their associated block storage devices      |
```

*NOTE:* Using the predefined templates the machine's memory size is automatically allocated based on the chosen shape 
and OCPU count.

The following Cluster API parameters are also available:

| Parameter                      | Default Value  | Description |
| ----------------------------   |----------------| ----------- |
| `CLUSTER_NAME`                 |                | The name of the workload cluster to create |
| `CONTROL_PLANE_MACHINE_COUNT`  | 1              | The number of control plane machines for the workload cluster.|

> | `KUBERNETES_VERSION`           |                | The Kubernetes version to use for the workload cluster. If unspecified, the value from OS environment variables or the .cluster-api/clusterctl.yaml config file will be used. |

Reviewer: ```suggestion
| `KUBERNETES_VERSION`           |                | The Kubernetes version installed on the workload cluster nodes. If this environement variable is not configured, the version must be specifed in the `.cluster-api/clusterctl.yaml` file |
```

The above suggestion is not correct if the assumption I made based on the format of the variable names that the table is defining the use of the each parameter when used as an environment variable is not true

Author: @shyamradhakrishnan might be a better person to answer this. Which is the same as `main` https://github.com/oracle/cluster-api-provider-oci/blob/main/docs/src/gs/create-workload-cluster.md?plain=1#L27

> | `NAMESPACE`                    |                | The namespace to use for the workload cluster. If unspecified, the current namespace will be used |

Reviewer: ```suggestion
| `NAMESPACE`                    |                | The namespace for the workload cluster. If not specified, the current namespace is used |
```

> | `POD_CIDR`                     | 192.168.0.0/16 | The CIDR range for the Kubernetes POD network. |

Reviewer: ```suggestion
| `POD_CIDR`                     | 192.168.0.0/16 | CIDR range of the Kubernetes pod-to-pod network |
```

> | `SERVICE_CIDR`                 | 10.128.0.0/12  | The CIDR for the Kubernetes services network.  |

Reviewer 3: This should be Calico's default service CIDR range

> | `SERVICE_CIDR`                 | 10.128.0.0/12  | The CIDR for the Kubernetes services network.  |

Reviewer: ```suggestion
| `SERVICE_CIDR`                 | 10.128.0.0/12  | CIDR range of the Kubernetes pod-to-services network  |
```

| `SERVICE_DOMAIN`               |                |  |
| `WORKER_MACHINE_COUNT`         |                | The number of worker machines for the workload cluster. |

## Create a new workload cluster on virtual instances using an Ubuntu custom image

> Run the command below to create a Kubernetes cluster with 1 control plane node and 1 worker node. This will setup 
> both the control plane and the worker nodes using the default information defined in the 
> [Workload Cluster Parameters](#workload-cluster-parameters) 
> table:

Reviewer: ```suggestion
The following command will create a workload cluster comprising a single 
control plane node and single worker node using the default values as specified in the preceeding 
[Workload Cluster Parameters](#workload-cluster-parameters) 
table:
```

The following Oracle Cloud Infrastructure (OCI) configuration parameters are available when creating a workload cluster on OCI:

| Parameter                                 | Default Value       | Description                                                                                                                                                                                                                                                                                                                                                                                                    |
|-------------------------------------------|---------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | `OCI_COMPARTMENT_ID`                      |                     | The OCID of the compartment where the OCI resources are to be created                                                                                                                                                                                                                                                                                                                                          |

Reviewer: "Are to be created" is a little cumbersome. Would "in which the resources are created" work better perhaps?

> | `OCI_IMAGE_ID`                          |                     | The OCID of the Compute Image (Oracle Linux or Ubuntu) with which to create the Kubernetes nodes                                                                                                                                                                                                                                                                                                                |

Author: - [ ] Update to be both control plane and worker images

> | `OCI_IMAGE_ID`                            |                     | The OCID of the Compute Image (Oracle Linux or <br/>Ubuntu) with which to create the Kubernetes nodes. This same image is used for both the control plane and the worker nodes.                                                                                                                                                                                                                                |

Reviewer: Neither compute nor image should be capitalised. Also, isn't this a custom image, not a compute image?

> | `OCI_IMAGE_ID`                            |                     | The OCID of the Compute Image (Oracle Linux or <br/>Ubuntu) with which to create the Kubernetes nodes. This same image is used for both the control plane and the worker nodes. Before picking an image look at the capabilities of the image.  Some images only support certain shapes, but most official images should support most (if not all) shapes. https://docs.oracle.com/en-us/iaas/api/#/en/iaas/20160918/Shape/ListShapes                                                 |

Author: Ah fair point. Thanks for that call out. I'll fix.

| `OCI_CONTROL_PLANE_SHAPE`                 | VM.Standard.E4.Flex | The shape of the Kubernetes nodes                                                                                                                                                                                                                                                                                                                                                                              |

> | `OCI_CONTROL_PLANE_SHAPE_MEMORY_IN_GBS`   |                     | The amount of memory in GBs to be allocated to the control plane instances. If not provided, the memory is automatically computed by compute API.                                                                                                                                                                                                                                                              |

Reviewer: I suggest expanding GBs to gigabytes unless you actually meant gibibytes. I would also replace "compute API" with something like "the memory is automatically allocated at 8GB per OCPU."

Author: > "the memory is automatically allocated at 8GB per OCPU."

but that isn't true. BM Machines have a much different GB to OPU allocation I believe.

Author: - [ ] After talking: "Memory is automatically allocated based on the chosen shape and OCPU count"

> | `OCI_CONTROL_PLANE_SHAPE_MEMORY_IN_GBS`   |                     | The amount of memory to be allocated to the instances. If not provided it is automatically computed by compute API.                                                                                                                                                                                                                                                                                                                                                                   |

Reviewer 3: Please replace with "The amount of memory in GBs to be allocated to the control plane instances. If not provided, the memory is automatically computed by compute API."

> | `OCI_CONTROL_PLANE_SHAPE_OCPUS`           | 1                   | The number of OCPUs allocated to the instance                                                                                                                                                                                                                                                                                                                                                                                                                                         |

Reviewer 3: Please replace with: "The number of OCPUs allocated to the control plane instance"

| `OCI_WORKER_SHAPE`                        | VM.Standard.E4.Flex | The shape of the Kubernetes worker nodes                                                                                                                                                                                                                                                                                                                                                                                                                                              |

> | `OCI_WORKER_SHAPE_MEMORY_IN_GBS`          |                     | The amount of memory to be allocated to the <br/>instances. If <br/>not provided it is automatically computed by compute API for VMs, but you will need to set this if you want to use Bare Metal instances.                                                                                                                                                                                                                                                                          |

Reviewer 3: Please replace with "The amount of memory in GBs to be allocated to the worker instances. If not provided, the memory is automatically computed by compute API."

> | `OCI_WORKER_SHAPE_OCPUS`                  | 1                   | The number of OCPUs allocated to the instance                                                                                                                                                                                                                                                                                                                                                                                                                                         |

Reviewer 3: Please replace with: "The number of OCPUs allocated to the worker instance"

| `OCI_SSH_KEY`                             |                     | The public SSH key to be added to the Kubernetes nodes. It can be used to login to the node and troubleshoot failures.                                                                                                                                                                                                                                                                                                                                                                |
| `OCI_CONTROL_PLANE_PV_TRANSIT_ENCRYPTION` | true                | [In-transit encryption](https://docs.oracle.<br/>com/en-us/iaas/Content/File/Tasks/intransitencryption.htm) provides a way to secure your data between instances and mounted file systems using TLS v.1.2 (Transport Layer Security) encryption. Only [some bare metal instances](https://docs.oracle.com/en-us/iaas/releasenotes/changes/60d602f5-abb3-4639-aa19-292a5744a808/) support In-transit encryption                                                                        |

> | `OCI_NODE_PV_TRANSIT_ENCRYPTION`          | true                | [In-transit encryption](https://docs.oracle.com/en-us/iaas/Content/File/Tasks/intransitencryption.htm) provides a way to secure your data between instances and mounted file systems using TLS v.1.2 (Transport Layer Security) encryption. Only [some bare metal instances](https://docs.oracle.com/en-us/iaas/releasenotes/changes/60d602f5-abb3-4639-aa19-292a5744a808/) support In-transit encryption                                                                             |

Reviewer 3: This should be `OCI_WORKER_PV_TRANSIT_ENCRYPTION`

> The following Cluster API parameters are also available:

Reviewer 3: Documentation error from previous commit: POD_CIDR cannot be 1.

Reviewer: Can we please replace "TLS v.1.2" with just "TLS 1.2" so that screen readers don't make it sound awful? Also, the "i" in in-transit encryption should be lowercase.

```bash
OCI_COMPARTMENT_ID=<compartment-id> \
OCI_IMAGE_ID=<ubuntu-custom-image-id> \

> OCI_CONTROL_PLANE_SHAPE=VM.Standard.E4.Flex \

Author: - [ ] Remove all of this and update docs to call out how to change the shape/ocpu/memory

OCI_CONTROL_PLANE_SHAPE_OCPUS=1 \
OCI_CONTROL_PLANE_SHAPE_MEMORY_IN_GBS= \

> OCI_NODE_SHAPE=VM.Standard.E4.Flex \

Author: - [ ] Change from `NODE` to `WORKER`

## Create a new workload cluster on bare metal instances using an Ubuntu custom image

Note the addition of the `OCI_CONTROL_PLANE_SHAPE` variables, `OCI_WORKER_SHAPE` variables to change the shape 
information from the `VM.Standard.E4.Flex` default. You will also need to set`OCI_PV_TRANSIT_ENCRYPTION=false` which is 
required for most BM shapes.

```bash
OCI_COMPARTMENT_ID=<compartment-id> \
OCI_IMAGE_ID=<ubuntu-custom-image-id> \
OCI_SSH_KEY=<ssh-key>  \
OCI_CONTROL_PLANE_SHAPE=BM.Standard2.52 \
OCI_CONTROL_PLANE_SHAPE_OCPUS=52 \
OCI_CONTROL_PLANE_SHAPE_MEMORY_IN_GBS= \
OCI_CONTROL_PLANE_PV_TRANSIT_ENCRYPTION=false \
OCI_WORKER_SHAPE=BM.Standard2.52 \
OCI_WORKER_SHAPE_OCPUS=52 \
OCI_WORKER_SHAPE_MEMORY_IN_GBS= \

> OCI_NODE_PV_TRANSIT_ENCRYPTION=false \

Reviewer 3: This should be `OCI_WORKER_PV_TRANSIT_ENCRYPTION`

## Create a new workload cluster on bare metal instances using an Ubuntu custom image

> Note the addition of the `OCI_CONTROL_PLANE_SHAPE` variables, `OCI_WORKER_SHAPE` variables to change the shape 

Reviewer 3: This needs to be reworded to something like:

You can specify different shapes for the control plane and worker nodes separately using the `OCI_CONTROL_PLANE_SHAPE` and  `OCI_WORKER_SHAPE` variables.

> information from the `VM.Standard.E4.Flex` default. You will also need to set`OCI_PV_TRANSIT_ENCRYPTION=false` which is 

Reviewer 3: Reword to:

If you are using bare metal shapes, you will also need to set `OCI_PV_TRANSIT_ENCRYPTION=false`

> Note the addition of the `OCI_CONTROL_PLANE_SHAPE` variables, `OCI_WORKER_SHAPE` variables to change the shape 
> information from the `VM.Standard.E4.Flex` default. You will also need to set`OCI_PV_TRANSIT_ENCRYPTION=false` which is 
> required for most BM shapes.

Reviewer: ```suggestion
The following command uses the `OCI_CONTROL_PLANE_SHAPE` and `OCI_WORKER_SHAPE` parameters to specify bare metal shapes instead of using CAPOCI's default virtual instance shape. The `OCI_CONTROL_PLANE_PV_TRANSIT_ENCRYPTION` and `OCI_WORKER_PV_TRANSIT_ENCRYPTION` parameters disable encryption of data in flight between the bare metal instance and the block storage resources.
```

The user experience with in-flight encryption is somewhat hostile: if set to true and the instance shape doesn't support it, we should just fail. That provides the best security and fast-fail result.

Author: > if set to true and the instance shape doesn't support it, we should just fail.

https://github.com/oracle/cluster-api-provider-oci/issues/34
